src/sentry/digests/utilities.py

```diff
@@ -23,3 +24,96 @@ def get_digest_metadata(digest):
                     end = record.datetime
 
     return start, end, counts
+
+
+def get_personalized_digests(project_id, digest, user_ids):
+    """
+    get_personalized_digests(project: Project, digest: Digest, users: Set[User]) -> Iterator[User, Digest]
+    """
+    # TODO(LB): I Know this is inefficent.
+    if not ProjectOwnership.objects.filter(project_id=project_id).exists():
+        return ProjectOwnership.Everyone
+    records = get_records_from_digests(digest)
+    events = [record.value.event for record in records]
+
+    events_by_actor = build_events_by_actor(project_id, events)
+    events_by_users = convert_actors_to_user_set(events_by_actor)
+
+    for user_id in user_ids:
+        yield user_id, build_custom_digest(digest, user_id, events_by_users)
+
+
+def get_records_from_digests(digest):
+    """
+    get_records_from_digests(digest: Digest) -> Set(Records)
+    """
+    records = set()
+    for rule, rule_groups in six.iteritems(digest):
+        for group, group_records in six.iteritems(rule_groups):
+            records.update(group_records)
+    return records
+
+
+def build_custom_digest(original_digest, user_id, events_by_users):
+    """
+    build_custom_digest(original_digest: Digest, user_id: Int, events_by_users: Map[User_Id:Set(Events)]) -> Digest
+    """
+    user_digest = OrderedDict()
+    for rule, rule_groups in six.iteritems(original_digest):
+        user_rule_groups = OrderedDict()
+        for group, group_records in six.iteritems(rule_groups):
+            user_group_records = [
+                record for record in group_records
+                if user_id in events_by_users[record.value.event]
+            ]
+            if user_group_records:
+                user_rule_groups[group] = user_group_records
+
+        if user_rule_groups:
+            user_digest[rule] = user_rule_groups
+    return user_digest
+
+
+def build_events_by_actor(project_id, events):
+    """
+    build_events_by_actor(project_id: Int, events: Set(Events)) -> Map[Actor:Set(Events)]
+    """
+    events_by_actor = {}
+    for event in events:
+        # TODO(LB): I Know this is inefficent. Just wanted to make as few changes
+        # as possible for now. Can follow up with another PR
+        actors = ProjectOwnership.get_owners(project_id, event)
```

Comment:
Be careful here, since this actually returns a tuple and I think you only want the first item: https://github.com/getsentry/sentry/blob/f869cb3a4419f2df2ada7aa7eefd85f8e8a7412c/src/sentry/models/projectownership.py#L63